akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -466,6 +466,91 @@ final class Broadcast[T](private val outputPorts: Int, eagerCancel: Boolean) ext
 
 }
 
+object Partition {
+
+  /**
+   * Create a new `Partition` stage with the specified input type.
+   *
+   * @param outputPorts number of output ports
+   * @param partitioner function deciding which output each element will be targeted
+   */
+  def apply[T](outputPorts: Int, partitioner: T ⇒ Int): Partition[T] = new Partition(outputPorts, partitioner)
+}
+
+/**
+ * Fan-out the stream to several streams. emitting an incoming upstream element to one downstream consumer according
+ * to the partitioner function applied to the element
+ *
+ * '''Emits when''' all of the outputs stops backpressuring and there is an input element available
+ *
+ * '''Backpressures when''' one of the outputs backpressure
+ *
+ * '''Completes when''' upstream completes
+ *
+ * '''Cancels when'''
+ *   when one of the downstreams cancel
+ */
+
+class Partition[T](outputPorts: Int, partitioner: T ⇒ Int) extends GraphStage[UniformFanOutShape[T, T]] {
+
+  val in: Inlet[T] = Inlet[T]("Partition.in")
+  val out: Seq[Outlet[T]] = Seq.tabulate(outputPorts)(i ⇒ Outlet[T]("Partition.out" + i))
+  override val shape: UniformFanOutShape[T, T] = UniformFanOutShape[T, T](in, out: _*)
+  private val outPulled = Array.fill[Boolean](outputPorts)(false)
+  private val outPending = mutable.Map.empty[Int, T]
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    var downstreamRunning = outputPorts
+
+    setHandler(in, new InHandler {
+      override def onPush() = {
+        val elem = grab(in)
+        val idx = partitioner(elem)
+        if (outPulled(idx)) {
+          if (!isClosed(out(idx))) {
+            push(out(idx), elem)
+            outPulled(idx) = false
+            if (outPulled.exists(x ⇒ x) && !hasBeenPulled(in))
+              pull(in)
+          }
+        } else {
+          outPending.put(idx, elem)
+        }
+      }
+    })
+
+    out.zipWithIndex.foreach {
+      case (o, idx) ⇒
+        setHandler(o, new OutHandler {
+          override def onPull() = {
+            outPending.get(idx) match {
+              case Some(elem) ⇒
+                if (!isClosed(o)) {
```

Comment:
I highly recommend the detailed state diagrams in the docs: http://doc.akka.io/docs/akka-stream-and-http-experimental/2.0-M2/scala/stream-customize.html#custom-processing-with-graphstage

They enumerate the roles of isAvailable, hasBeenPulled, isClosed queries, and all available internal and external state transitions. It is fully tested and can be considered as a full specification of the port API.